### ISSUE TITLE
"The device was not grabbed" when remapping BTN_MIDDLE to BTN_LEFT

### ISSUE THREAD

**Issue by user529481** [sezanzeb/input-remapper#1022]
Hello, I'm trying to remap the left button of my mouse (Evoluent VerticalMouse D) to the middle button for my arthritis. I'm getting "The device was not grabbed error" when I click "Apply". 

The program did allow me to register both the input and output correctly. I was even able to map the middle mouse button to a letter on the keypad successfully. Similarly I was able to map keyboard keys to other keys just fine. 

When I run `xinput list`
```
WARNING: running xinput against an Xwayland server. See the xinput man page for details.
⎡ Virtual core pointer                     id=2 [master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer               id=4 [slave  pointer  (2)]
⎜   ↳ xwayland-pointer:15                     id=6 [slave  pointer  (2)]
⎜   ↳ xwayland-relative-pointer:15             id=7 [slave  pointer  (2)]
⎜   ↳ xwayland-pointer-gestures:15             id=8 [slave  pointer  (2)]
⎣ Virtual core keyboard                   id=3 [master keyboard (2)]
    ↳ Virtual core XTEST keyboard             id=5 [slave  keyboard (3)]
    ↳ xwayland-keyboard:15                     id=9 [slave  keyboard (3)]
```

**Comment by sezanzeb** [comment]
Before you hit "Apply", is there some other error shown on the bottom of the window?

**Comment by Zoldronesk** [comment]
Getting the exact same issue with a Logitech G502.

System:
  Kernel: 6.8.0-51-generic arch: x86_64 bits: 64 compiler: gcc v: 13.3.0 clocksource: tsc
  Desktop: Cinnamon v: 6.2.9 tk: GTK v: 3.24.41 wm: Muffin v: 6.2.0 vt: 7 dm: LightDM v: 1.30.0
    Distro: Linux Mint 22 Wilma base: Ubuntu 24.04 noble
Machine:
  Type: Desktop Mobo: INTEL model: B75 serial: <superuser required> uuid: <superuser required>
    UEFI: American Megatrends v: 4.6.5 date: 06/03/2021
CPU:
  Info: quad core model: Intel Core i5-3470 bits: 64 type: MCP smt: <unsupported> arch: Ivy Bridge
    rev: 9 cache: L1: 256 KiB L2: 1024 KiB L3: 6 MiB
  Speed (MHz): avg: 2018 high: 2785 min/max: 1600/3600 cores: 1: 1596 2: 2785 3: 1597 4: 2096
    bogomips: 25543
  Flags: avx ht lm nx pae sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx
Graphics:
  Device-1: NVIDIA GK104 [GeForce GTX 660 Ti] vendor: eVga.com. driver: nvidia v: 470.256.02
    arch: Kepler pcie: speed: 2.5 GT/s lanes: 8 bus-ID: 01:00.0 chip-ID: 10de:1183 class-ID: 0300
  Display: x11 server: X.Org v: 21.1.11 with: Xwayland v: 23.2.6 driver: X: loaded: nouveau
    unloaded: fbdev,modesetting,vesa failed: nvidia gpu: nvidia display-ID: :0 screens: 1
  Screen-1: 0 s-res: 1920x1080 s-dpi: 101 s-size: 483x272mm (19.02x10.71") s-diag: 554mm (21.82")
  Monitor-1: Unknown-1 mapped: DVI-D-0 res: 1920x1080 hz: 60 dpi: 102
    size: 477x268mm (18.78x10.55") modes: 640x480
  API: EGL v: 1.5 hw: drv: nvidia platforms: device: 0 drv: nvidia device: 2 drv: swrast gbm:
    drv: kms_swrast surfaceless: drv: swrast x11: drv: nvidia inactive: wayland,device-1
  API: OpenGL v: 4.6.0 compat-v: 4.5 vendor: nvidia mesa v: 470.256.02 glx-v: 1.4
    direct-render: yes renderer: NVIDIA GeForce GTX 660 Ti/PCIe/SSE2
  API: Vulkan v: 1.3.275 layers: 4 surfaces: xcb,xlib device: 0 type: discrete-gpu driver: N/A
    device-ID: 10de:1183 device: 1 type: cpu driver: N/A device-ID: 10005:0000
Audio:
  Device-1: Intel 7 Series/C216 Family High Definition Audio driver: snd_hda_intel v: kernel
    bus-ID: 00:1b.0 chip-ID: 8086:1e20 class-ID: 0403
  Device-2: NVIDIA GK104 HDMI Audio vendor: eVga.com. driver: snd_hda_intel v: kernel pcie:
    speed: 8 GT/s lanes: 8 bus-ID: 01:00.1 chip-ID: 10de:0e0a class-ID: 0403
  API: ALSA v: k6.8.0-51-generic status: kernel-api
  Server-1: PipeWire v: 1.0.5 status: active with: 1: pipewire-pulse status: active
    2: wireplumber status: active 3: pipewire-alsa type: plugin
Network:
  Device-1: Realtek RTL810xE PCI Express Fast Ethernet driver: r8169 v: kernel pcie:
    speed: 2.5 GT/s lanes: 1 port: d000 bus-ID: 03:00.0 chip-ID: 10ec:8136 class-ID: 0200
  IF: enp3s0 state: down mac: <filter>
  Device-2: Realtek RTL8188EUS 802.11n Wireless Network Adapter driver: rtl8xxxu type: USB
    rev: 2.0 speed: 480 Mb/s lanes: 1 bus-ID: 1-1.4:3 chip-ID: 0bda:8179 class-ID: 0000
    serial: <filter>
  IF: wlx3c52a1daa812 state: down mac: <filter>
  IF-ID-1: enx06377489e1f2 state: unknown speed: -1 duplex: half mac: <filter>
  IF-ID-2: wg0-mullvad state: unknown speed: N/A duplex: N/A mac: N/A
Drives:
  Local Storage: total: 912.89 GiB used: 201.5 GiB (22.1%)
  ID-1: /dev/sda vendor: Kingston model: SA400S37480G size: 447.13 GiB speed: 6.0 Gb/s tech: SSD
    serial: <filter> fw-rev: T1A3 scheme: GPT
  ID-2: /dev/sdb vendor: Western Digital model: WD5000AAKX-001CA0 size: 465.76 GiB
    speed: 3.0 Gb/s tech: N/A serial: <filter> fw-rev: 1H15 scheme: MBR
Partition:
  ID-1: / size: 435.01 GiB used: 201.21 GiB (46.3%) fs: ext4 dev: /dev/dm-1 mapped: vgmint-root
  ID-2: /boot size: 1.61 GiB used: 289.7 MiB (17.6%) fs: ext4 dev: /dev/sda3
  ID-3: /boot/efi size: 512 MiB used: 6.1 MiB (1.2%) fs: vfat dev: /dev/sda2
Swap:
  ID-1: swap-1 type: partition size: 1.91 GiB used: 0 KiB (0.0%) priority: -2 dev: /dev/dm-2
    mapped: vgmint-swap_1
USB:
  Hub-1: 1-0:1 info: full speed or root hub ports: 2 rev: 2.0 speed: 480 Mb/s lanes: 1
    chip-ID: 1d6b:0002 class-ID: 0900
  Hub-2: 1-1:2 info: Intel Integrated Rate Matching Hub ports: 6 rev: 2.0 speed: 480 Mb/s
    lanes: 1 chip-ID: 8087:0024 class-ID: 0900
  Device-1: 1-1.4:3 info: Realtek RTL8188EUS 802.11n Wireless Network Adapter type: Network
    driver: rtl8xxxu interfaces: 1 rev: 2.0 speed: 480 Mb/s lanes: 1 power: 500mA chip-ID: 0bda:8179
    class-ID: 0000 serial: <filter>
  Hub-3: 2-0:1 info: full speed or root hub ports: 2 rev: 2.0 speed: 480 Mb/s lanes: 1
    chip-ID: 1d6b:0002 class-ID: 0900
  Hub-4: 2-1:2 info: Intel Integrated Rate Matching Hub ports: 8 rev: 2.0 speed: 480 Mb/s
    lanes: 1 chip-ID: 8087:0024 class-ID: 0900
  Hub-5: 3-0:1 info: hi-speed hub with single TT ports: 4 rev: 2.0 speed: 480 Mb/s lanes: 1
    chip-ID: 1d6b:0002 class-ID: 0900
  Device-1: 3-1:2 info: Motorola PCS moto g 5G - 2023 type: CDC-data driver: rndis_host
    interfaces: 2 rev: 2.0 speed: 12 Mb/s lanes: 1 power: 500mA chip-ID: 22b8:2e24 class-ID: 0a00
    serial: <filter>
  Device-2: 3-3:3 info: Logitech G502 SE HERO Gaming Mouse type: mouse,HID
    driver: hid-generic,usbhid interfaces: 2 rev: 2.0 speed: 12 Mb/s lanes: 1 power: 300mA
    chip-ID: 046d:c08b class-ID: 0300 serial: <filter>
  Device-3: 3-4:4 info: Elan Micro ActiveJet K-2024 Multimedia Keyboard type: keyboard,HID
    driver: hid-generic,usbhid interfaces: 2 rev: 2.0 speed: 1.5 Mb/s lanes: 1 power: 100mA
    chip-ID: 04f3:0103 class-ID: 0300
  Hub-6: 4-0:1 info: super-speed hub ports: 4 rev: 3.0 speed: 5 Gb/s lanes: 1 chip-ID: 1d6b:0003
    class-ID: 0900
Sensors:
  System Temperatures: cpu: 26.0 C mobo: N/A gpu: nvidia temp: 39 C
  Fan Speeds (rpm): N/A gpu: nvidia fan: 30%
Repos:
  Packages: 2436 pm: dpkg pkgs: 2406 pm: flatpak pkgs: 30
  No active apt repos in: /etc/apt/sources.list
  Active apt repos in: /etc/apt/sources.list.d/mullvad.list
    1: deb [signed-by=/usr/share/keyrings/mullvad-keyring.asc arch=amd64] https: //repository.mullvad.net/deb/beta wilma main
  Active apt repos in: /etc/apt/sources.list.d/official-package-repositories.list
    1: deb http: //packages.linuxmint.com wilma main upstream import backport
    2: deb http: //archive.ubuntu.com/ubuntu noble main restricted universe multiverse
    3: deb http: //archive.ubuntu.com/ubuntu noble-updates main restricted universe multiverse
    4: deb http: //archive.ubuntu.com/ubuntu noble-backports main restricted universe multiverse
    5: deb http: //security.ubuntu.com/ubuntu/ noble-security main restricted universe multiverse
Info:
  Memory: total: 16 GiB available: 15.57 GiB used: 2.59 GiB (16.6%)
  Processes: 268 Power: uptime: 19m states: freeze,standby,mem,disk suspend: s2idle wakeups: 0
    hibernate: platform Init: systemd v: 255 target: graphical (5) default: graphical
  Compilers: gcc: 13.3.0 Client: Cinnamon v: 6.2.9 inxi: 3.3.34

**Comment by sezanzeb** [comment]
Hm. Could you please share a screenshot of the preset?

**Comment by sezanzeb** [comment]
And the output of `input-remapper-control --version` please.

**Comment by Zoldronesk** [comment]
_input-remapper 2.0.1  https://github.com/sezanzeb/input-remapper
python-evdev 1.7.0_



...but you'll be glad to know I was simply being an idiot and assuming that pressing the key I wanted to map would record it instead of just letting me know what to type in the field.

**Comment by sezanzeb** [comment]
I pushed changes to the main branch about the error messages. I hope it is clearer now. When you hit apply, it will show validation errors now if it fails.

**Comment by alangomezhuerta** [comment]
Hi !,  I am trying like crzy to remap Control L to a macro ,   what I would like to achieve is :  Control_L to Escape on single press alone, Control_L on press and hold for more than 250 milliseconds or combined to other keys.
Somehow trying to understand syntax I am supposing I should inject something like : 
if_tap(key(KEY_ESC), mod_tap(Control_L, Control_L, 250), 250)
Though I'm stuck in errors :+1: 
![input_remapper](https://github.com/user-attachments/assets/30e24c59-64bf-4b0f-996a-8197d2d25b62)

input-remapper 2.0.0 9e04df79ca4dca77516a9864d740b9b2d4aef788 https://github.com/sezanzeb/input-remapper
python-evdev 1.4.0

**Comment by sezanzeb** [comment]
This is a different issue, please open a new one next time.

You need to install the very latest version from github for mod_tap to work. mod_tap will be released properly in 2.1.0.

Also, I think you probably just want `mod_tap(KEY_ESC, Control_L, 250)`, or, depending on what feels right, `if_single(key(KEY_ESC), key(Control_L), 250)`

**Comment by jpeletier** [comment]
I fell for the same and ended up in this thread. I thought what I typed was what was configured, but thanks to this thread I understood that I was supposed to look at the code and type it myself.

In any case, thanks for writing this tool!
